### PR TITLE
Add XDCAnnotation type hints, end-to-end test, empty arg list support

### PR DESCRIPTION
**Pull request by fabianschuiki**
This commit does three things:

- Add a `CustomAnnotations` test to the midasexamples test suite which is basically the shift register test with additional uses of the `XDC(...)` Chisel API. The test checks whether the expected files get generated and contain the expected lines.

- Add an early out to `WriteXDCFile.formatArguments` in case the format argument list is empty. This allows the user to specify an XDC constraint that contains no format arguments, which until now would cause an exception in `argumentList.head`. The end-to-end test checks for this.

- Add type hints to the `XDCAnnotation` such that the `XDCDestinationFile` case objects get serialized properly.

#### Related PRs / Issues

None

#### UI / API Impact

No impact

#### Verilog / AGFI Compatibility

No impact

### Contributor Checklist
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [ ] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [ ] If applicable, did you regenerate and publicly share default AGFIs?

- [ ] If applicable, did you apply the `ci:fpga-deploy` label?

- [ ] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
Note: to run CI on PRs from forks, comment `@Mergifyio copy main` and manage the change from the new PR.
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: sim/src/test/scala/midasexamples/TutorialSuite.scala [hunk 1]
```diff
@@ -377,6 +377,23 @@ class TerminationF1Test extends TutorialSuite("TerminationModule") {
...
 
+class CustomConstraintsF1Test extends TutorialSuite("CustomConstraints") {
+  def readLines(filename: String): List[String] = {
+    val file = new File(genDir, s"/${filename}")
+    Source.fromFile(file).getLines.toList
+  }
+  it should s"generate synthesis XDC file" in {
+    val xdc = readLines("FireSim-generated.synthesis.xdc")
+    xdc should contain("constrain_synth1")
+    xdc should contain("constrain_synth2 [reg firesim_top/top/sim/target/CustomConstraints_/dut/r0]")
```
Comment (davidbiancolin) [inline sim/src/test/scala/midasexamples/TutorialSuite.scala:388]: Do you have any ideas for ways we could avoid hardcoding a module path here? These are liable to change.

File: sim/src/test/scala/midasexamples/TutorialSuite.scala [hunk 2]
```diff
@@ -377,6 +377,23 @@ class TerminationF1Test extends TutorialSuite("TerminationModule") {
...
+  }
+  it should s"generate implementation XDC file" in {
+    val xdc = readLines("FireSim-generated.implementation.xdc")
+    xdc should contain("constrain_impl1")
+    xdc should contain("constrain_impl2 [reg WRAPPER_INST/CL/firesim_top/top/sim/target/CustomConstraints_/dut/r1]")
+  }
+}
+
 // Suite Collections
 class ChiselExampleDesigns extends Suites(
```
Comment (davidbiancolin) [inline sim/src/test/scala/midasexamples/TutorialSuite.scala:398]: You should add your constraint test to one of the CI groups below.

**Review by davidbiancolin — COMMENTED** [review]
This is a nice improvement, thanks for fixing the 0-argument case. 

There are a lot of hardcoded strings in the test, but this is consistent with what i've done elsewhere.

File: sim/src/test/scala/midasexamples/TutorialSuite.scala (hunk 1 above)
Comment (fabianschuiki) [reply sim/src/test/scala/midasexamples/TutorialSuite.scala:388]: Yeah the hardcoded strings also make me sad.

Theoretically you could try to throw a custom annotation into the input and then check what the target of the anno got transformed into at the end of the flow, and then check whether that also shows up here in the XDC output file. Alternatively, some form of metadata could be useful, like the ability to assign a tag to something in the IR and then at the end of the compilation asking the compiler to dig up the concrete final path that tag landed on. But that's basically what the annotations already do.

File: sim/src/main/scala/midasexamples/CustomConstraints.scala [hunk 3]
```diff
@@ -0,0 +1,25 @@
+//See LICENSE for license details.
+
+package firesim.midasexamples
+
+import chisel3._
+import freechips.rocketchip.config.Parameters
+import midas.targetutils.xdc._
+
+class CustomConstraintsDUT extends Module {
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply sim/src/main/scala/midasexamples/CustomConstraints.scala:9]: Yeah good point. Let me do that.

[the remark this answers is not in the document]
Comment (fabianschuiki) [reply sim/src/main/scala/midasexamples/CustomConstraints.scala]: Done

File: sim/src/test/scala/midasexamples/TutorialSuite.scala (hunk 2 above)
Comment (fabianschuiki) [reply sim/src/test/scala/midasexamples/TutorialSuite.scala:398]: Done

**Comment by fabianschuiki** [comment]
Applied your suggestions @davidbiancolin, thanks a lot :+1:

File: sim/src/test/scala/midasexamples/TutorialSuite.scala (hunk 1 above)
Comment (davidbiancolin) [reply sim/src/test/scala/midasexamples/TutorialSuite.scala:388]: Yeah... While not perfect, I think both of those ideas are pretty reasonable. 

In the short term you can probably just wildcard the middle component of the path? The pass-level test in GoldenGate can check for complete path correctness.

Comment (fabianschuiki) [reply sim/src/test/scala/midasexamples/TutorialSuite.scala:388]: Yeah I like that approach. We only care about the prefix and suffix of the path. Let me tweak this.

Comment (fabianschuiki) [reply sim/src/test/scala/midasexamples/TutorialSuite.scala]: Finally figured out how to do this. So many DSLs :see_no_evil:

**Review by davidbiancolin — COMMENTED** [review]
This looks good to me, but need to hold off merging until Sagar cuts a release for the FireSim tutorial.

**Comment by fabianschuiki** [comment]
Sounds great, thanks for the reviews and suggestions :tada:!

**Review by davidbiancolin — APPROVED** [review]
